cpp/src/Slice/SliceUtil.cpp

```diff
@@ -788,3 +788,72 @@ bool Slice::opCompressReturn(const OperationPtr& op)
 {
     return opCompress(op, false);
 }
+
+string Slice::getDeprecateReason(const ContainedPtr& p, bool checkContainer)
+{
+    // Check if there was `deprecate` metadata directly on the entity.
+    auto deprecateMsg = p->findMetadata("deprecate");
+    // If the entity wasn't directly deprecated, check if it's container is.
+    if (!deprecateMsg && checkContainer)
+    {
+        if (ContainedPtr p2 = ContainedPtr::dynamicCast(p->container()))
+        {
+            deprecateMsg = p2->findMetadata("deprecate");
+        }
+    }
+
+    if (deprecateMsg)
+    {
+        if (deprecateMsg->empty())
+        {
+            return "This " + p->kindOf() + " has been deprecated";
+        }
+        return *deprecateMsg;
+    }
+    return "";
+}
+
+pair<string, string> Slice::parseMetadata(const string& metadata)
+{
+    // Check if the metadata has any arguments.
+    auto argumentStart = metadata.find("(");
+    if (argumentStart == string::npos)
+    {
+        return make_pair(metadata, "");
+    }
+    else
+    {
+        // Make sure the parentheses are balanced. For metadata with arguments, the final character must be ')'.
+        auto argumentEnd = metadata.length() - 1;
+        if (metadata.at(argumentEnd) != ')')
+        {
+            unit->error("missing closing parentheses for metadata arguments");
+        }
+        return make_pair(metadata.substr(0, argumentStart), metadata.substr(argumentStart + 1, argumentEnd));
+    }
+}
+
+map<string, string> Slice::parseMetadata(const StringList& metadata)
+{
+    map<string, string> parsedMetadata;
+    for (const auto& m : metadata)
+    {
+        parsedMetadata.insert(parseMetadata(m));
+    }
+    return parsedMetadata;
+}
+
+bool Slice::hasMetadata(const string& directive, const map<string, string>& metadata)
+{
+    return metadata.find(directive) != metadata.end();
+}
+
+optional<string> Slice::findMetadata(const string& directive, const map<string, string>& metadata)
+{
+    auto match = metadata.find(directive);
+    if (match != metadata.end())
+    {
+        return match->second;
+    }
+    return nullopt;
```

Comment:
Can we use the ternary operator for these simple conditions?

```suggestion
    return match != metadata.end() ? match->second : nullopt;
```

Reply:
Yes, absolutely!